wallet, rpc: deprecate settxfee and paytxfee

polespinasa opened the pull request:
**Summary**

This PR deprecates the settxfee RPC and paytxfee setting, marking it for removal in Bitcoin Core 31.0.

**Motivation**

The PR was initially motivated by https://github.com/bitcoin/bitcoin/issues/31088. The intention was to create a new function `settxfeerate` to allow users to set a static fee rate in `sat/vB` instead of `btc/kvB`.

The `settxfee` RPC allows users to set a static fee rate for all transactions created by the wallet. However, in a dynamic fee environment, this can lead to poor fee choices, either overpaying when the mempool is empty or underpaying when congestion is high. The preferred approach is to rely on fee estimation, which is designed to adapt to network conditions, and is the one by default. Same argument apply for `paytxfee` setting.

During discussion the consensus was that static fee settings are a footgun and that users should instead specify the fee rate per transaction if they don't want to rely on the fee estimation. Given this, rather than introducing a `settxfeerate` alternative, this PR goes towards removing `settxfee` and `paytxfee` entirely.

**Key Changes**

`settxfee` and `paytxfee` is now deprecated and will be removed in Bitcoin Core 31.0.
Users should rely on fee estimation or explicitly specify a fee rate when constructing transactions.

**Impact on Users**

If users currently use settxfee or paytxfee, they should transition to specifying fees per transaction.
No immediate breakage in 30.0 (must use `-deprecatedrpc=settxfee`), but `settxfee` and `paytxfee` will be removed in 31.0.

**Alternative Approaches Considered**

A settxfeerate alternative (using sat/vB) was initially proposed but ultimately rejected in favor of deprecating static fee setting entirely.

**Notes for removal**
- When removing paytxfee we should also update txconfirmtarget startup option help text.
- Get back the comment from `rpc_deprecated.py` test. [+info](https://github.com/bitcoin/bitcoin/pull/31278#discussion_r1998876768)

polespinasa commented:
The functional test runs correctly on my local environment:
``64/313 - rpc_settxfeerate.py passed, Duration: 1 s``

I don't know if anything needs to be done in the IC to run the tests correctly.
Help regarding this is greatly appreciated.

fjahr commented on test/functional/rpc_settxfeerate.py:
```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env python3
+# Copyright (c) 2014-2022 The Bitcoin Core developers
```
Why 2014-2022?

fjahr commented on src/wallet/rpc/spend.cpp:
```diff
@@ -456,6 +457,53 @@ RPCHelpMan settxfee()
+                    RPCResult::Type::BOOL, "", "Returns true if successful"
+                },
+                RPCExamples{
+                    HelpExampleCli("settxfee", "1")
```
Examples copy+pasted but not updated

polespinasa replied:
thanks for the observation, will correct it

fjahr commented on src/wallet/rpc/spend.cpp:
```diff
@@ -418,7 +418,8 @@ RPCHelpMan sendmany()
 {
     return RPCHelpMan{"settxfee",
-                "\nSet the transaction fee rate in " + CURRENCY_UNIT + "/kvB for this wallet. Overrides the global -paytxfee command line parameter.\n"
+                "\nWill be deprecated see 'settxfeerate'.\n"
```
Why "will be deprecated"? Why not actually deprecate it?

polespinasa replied:
Avoid incompatibility problems with programs actually using this RPC call.

fjahr replied:
That's what we have the `-deprecatedrpc` parameter for

polespinasa replied:
I will mark this as resolved as we're discussing the same in the other comment thread

fjahr commented on src/wallet/rpc/wallet.cpp:
```diff
@@ -1156,7 +1157,8 @@ Span<const CRPCCommand> GetWalletRPCCommands()
         {"wallet", &sethdseed},
         {"wallet", &setlabel},
-        {"wallet", &settxfee},
+        {"hidden", &settxfee},
```
Do we set deprecated rpc methods to hidden? I'm not sure I have seen that before.

polespinasa replied:
it's what it's suggested on https://github.com/bitcoin/bitcoin/issues/31088

fjahr replied:
Looking back it seems like we haven't hidden them, just deprecated. That's also what I would suggest here.

polespinasa replied:
Do you have an example of deprecated RPC so I can compare on how it was done before?

fjahr replied:
Checking a few of these should give you an overview: https://github.com/bitcoin/bitcoin/pulls?q=is%3Apr+Deprecate+in%3Atitle+rpc+in%3Atitle+is%3Aclosed+

polespinasa replied:
Thanks!
Applied following some of the examples.

murchandamus replied:
It seems to me that you agreed with @fjahr on his comment here, but the RPC is still being hidden. Is that an oversight?

polespinasa replied:
> @fjahr @ismaelsadeeq `txfeerate` is now deprecated, but I kept it hidden. IMO it makes sense to have a deprecated rpc call hidden as only users who were using it before should know about it and may need info, they can continue using `help settxfee`

I justified it in another comment below. If you consider that it should not be hidden for some reason or if it is the way it has always been done and you want to keep the form, I will change.

ismaelsadeeq replied:
Makes sense to me to be hidden but no strong opinion

fjahr replied:
I posted the link above with example and I haven't seen a case where the RPC was hidden in there, so I think "the way it has always been" was to not deprecate. We haven't removed a whole RPC in a while though. But we can look at the options and returned values that were deprecated too: In that case it would make sense to remove the documentation of these options at the same time as we are deprecating them if we are following the justification for hiding. But instead we only add information that the option is deprecated and keep the documentation in place. My vote would still to not hide since it's a form of documentation that may still be valuable but I am not feeling so strongly about it that it's a blocker. It's an interesting question in general though, so I think I will as in IRC if others have an opinion on this. It would be good to have clear guidance on this.

IMO it is considered best practice to retain documentation, generally speaking but it's just a feeling from having seen thousands of documentation pages with deprecation notices at the top. There seem to be some project going in the other direction as well thought, a quick search revealed this for example: https://discourse.gohugo.io/t/documentation-for-deprecated-functions-methods-etc/46919

jonatack replied:
I don't recall deprecated RPCs being set to hidden in the past. Generally we do that for developer-only ones that could represent footguns for users. If this RPC is considered to be sufficiently dangerous, that could perhaps justify hiding it, but not due only to being deprecated.

ismaelsadeeq replied:
My thinking of deprecating a feature is discouraging its use without outright disabling it, only allowing it when explicitly required.

Hiding it from the help text is a good step in this direction.

You can still access details about the RPC by running:

bitcoin-cli settxfee --help

You just won’t see the feature description in the overview of available RPCs.

As for paytxfee, it is still listed in the startup option help text.

AFAICT, it is not deprecated it's just warnings that are are emitted when it is used. See:
https://github.com/bitcoin/bitcoin/pull/31278#discussion_r1977554732

jonatack replied:
Software clients that use the deprecated RPC API will see breakage when they update and look for why. It may be more confusing to no longer see the RPC in the list and potentially think it has been removed outright ("core devs broke my app without warning") rather than deprecated behind a flag. Hopefully, people read the release notes but 🤷.

ismaelsadeeq replied:
fair point cc @polespinasa

polespinasa replied:
> It may be more confusing to no longer see the RPC in the list and potentially think it has been removed.

That's a fair point and a good reason to not hide it. I will just revert that. 

> It would be good to have clear guidance on this.

Would be good to have a written procedure on the dev notes to follow always the same procedure in the future and avoid repeating the discussion.

ismaelsadeeq replied:
> Would be good to have a written procedure on the dev notes to follow always the same procedure in the future and avoid repeating the discussion.

Good idea for a follow up PR, a section on removing a feature will be nice.

But we have different feature category.

1. RPC
2. Startup option
3. Rest
4. ZMQ interface
5. Wallet settings
6. In the future ( multiprocess interface)

We should document how we remove each. 

Documenting for RPC is a good first step.

polespinasa replied:
Left an open issue for it: https://github.com/bitcoin/bitcoin/issues/31980
I think @jonatack and @fjahr have more to say here than me.

polespinasa replied:
> Good idea for a follow up PR, a section on removing a feature will be nice.

@ismaelsadeeq I think you might want to take a look at this PR https://github.com/bitcoin/bitcoin/pull/30142

fjahr commented on test/functional/rpc_settxfeerate.py:
```diff
@@ -0,0 +1,56 @@
+# Copyright (c) 2014-2022 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test the wallet accounts properly when there are cloned transactions with malleated scriptsigs."""
```
I don't think that's the right description

polespinasa replied:
copy paste for a template, I will update it

fjahr reviewed (commented):
Did you try to follow the approach suggested here: https://github.com/bitcoin/bitcoin/issues/31088#issuecomment-2413394142? If not, can you say why it didn't work?

polespinasa commented:
> Did you try to follow the approach suggested here: [#31088 (comment)](https://github.com/bitcoin/bitcoin/issues/31088#issuecomment-2413394142)? If not, can you say why it didn't work?

I would have to look deeper on how to do that.
But I don't see why it would be better than this approach. Deprecating here is easy and there will not be the need to modify code other than deleting the actual function.

ismaelsadeeq reviewed (commented):
Nice fixing the ambiguity.

But generally, I’m Concept ~0 on setting static feerate for wallet transaction as a whole because of the current dynamic fee environment, it’s just a footgun.
It will make a transaction either pay too high or too low.

Feerate estimate from fee estimation algorithm should be used in most cases, and when that algorithm is deem inaccurate by users they should pass the fee rate for each individual transaction.

We should lean towards deprecating this RPC instead.

polespinasa commented on src/wallet/rpc/spend.cpp:
```diff
@@ -418,10 +418,11 @@ RPCHelpMan sendmany()
                 "Can be deactivated by passing 0 as the fee. In that case automatic fee selection will be used by default.\n",
                 {
-                    {"amount", RPCArg::Type::AMOUNT, RPCArg::Optional::NO, "The transaction fee rate in " + CURRENCY_ATOM + "/vB"},
+                    {"amount", RPCArg::Type::AMOUNT, RPCArg::Optional::NO, "The transaction fee rate in " + CURRENCY_UNIT + "/kvB"},
```
That was a mistake in the previous commit, as it's the ``settxfee`` function and not ``settxfeerate``.

polespinasa commented:
> Feerate estimate from fee estimation algorithm should be used in most cases, and when that algorithm is deem inaccurate by users should they should pass the fee rate for each individual transaction.

I agree that the algorithm should be used in most cases, and actually that's how it's working according to the RPC call definition. By default it uses the dynamic fee so this is only an extra option to set a static fee in case anyone needs too.
https://github.com/bitcoin/bitcoin/blob/b96cfae733a8a8eb79ee894a7cbe903fa955bcec/src/wallet/rpc/spend.cpp#L422

polespinasa commented on test/functional/rpc_settxfeerate.py:
```diff
@@ -0,0 +1,50 @@
+
+    	self.log.debug("Test that settxfee limit (10000 sat/vB) can not be surpassed")
+    	assert_raises_rpc_error(-8, "txfee cannot be more than wallet max tx fee (0.10000000 BTC/kvB)", node.settxfeerate, 10001)
+
```
Good point, added!

fjahr commented on test/functional/rpc_settxfeerate.py:
```diff
@@ -1,9 +1,10 @@
 #!/usr/bin/env python3
-# Copyright (c) 2014-2022 The Bitcoin Core developers
+# Copyright (c) 2014-2024 The Bitcoin Core developers
```
This file didn't exist before 2024, so it should just say 2024

polespinasa commented:
@fjahr @ismaelsadeeq ``txfeerate`` is now deprecated, but I kept it hidden. IMO it makes sense to have a deprecated rpc call hidden as only users who were using it before should know about it and may need info, they can continue using ``help settxfee``

polespinasa commented:
I'm not sure why that CI check ,``Win64 native, VS2022 (pull_request)``, fails. Help is appreciated here.

Prabhat1308 commented on test/functional/rpc_settxfeerate.py:
```diff
@@ -0,0 +1,83 @@
+        assert_equal(node.settxfee(0.001), True)
+        assert_equal(str(node.getwalletinfo()['paytxfee']), "0.00100000")
+
+        self.log.debug("Test that settxfeerate set the feerate in v/vB")
```
```suggestion
        self.log.debug("Test that settxfeerate set the feerate in sat/vB")
```

polespinasa replied:
Thanks!

luke-jr commented:
See also #20391 for previous work in this area

ismaelsadeeq commented on src/wallet/rpc/spend.cpp:
```diff
@@ -437,6 +437,10 @@ RPCHelpMan settxfee()
     LOCK(pwallet->cs_wallet);
 
+    if (!pwallet->chain().rpcEnableDeprecated("settxfee")) {
+        throw JSONRPCError(RPC_METHOD_DEPRECATED, "Using settxfee is deprecated. Use settxfeerate instead or restart bitcoind with -deprecatedrpc=settxfee. This functionality will be removed in x.xx");
```
I think just saying it is deprecated is enough, because you are giving half baked information by not specifying when it will be removed!

fjahr replied:
How is that different from just saying it is deprecated without giving information when it will be removed?  Either way the information is half-baked.

polespinasa replied:
Shall we maybe agree on when it can be deprecated? :)

ismaelsadeeq replied:
@polespinasa 

Resurfacing an IRC discussion about this.

> 11:39 < abubakarsadiq> I've searched but did not see guidelines on how long will a deprecated feature (e.g) RPC will take for it to be removed! I've seen an example from d4cdbd6fb6ac3663d069307c4fd0078f4ecf0245 that the RPC will be removed in the next release. Is that the case? if so @tdb3 maybe document that as well in #30142 
11:39 <@gribble> https://github.com/bitcoin/bitcoin/issues/30142 | doc: add guidance for RPC to developer notes by tdb3 · Pull Request #30142 · bitcoin/bitcoin · GitHub
11:40 -!- saturday7 [~saturday7@59.167.129.22] has quit [Ping timeout: 248 seconds]
11:44 < sipa> abubakarsadiq: generally, one major release (so marked deprecated in version N.0, removed in (N+1).0)
11:44 < sipa> but of course, because it requires actual work (a PR being written and reviewed) during the N+1 release cycle, that may not happen, either because it's forgotten, or because people worry that too many users rely on it still
11:47 < abubakarsadiq> Thanks so it's a case by case basis, I see the reason why having an estimate maybe an issue

I think you can say that it is deprecated in 29.0 and will be removed in 30.0

ismaelsadeeq commented:
@murchandamus pinging you here , since this PR originates from our discussion here https://github.com/bitcoin/bitcoin/pull/29278#discussion_r1526664705.

See my comment above https://github.com/bitcoin/bitcoin/pull/31278#pullrequestreview-2430770541.

I think it would be better to simply deprecate`settxfee` and not introduce `settxfeerate`.
 In future releases, we should even consider removing support for `settxfee` entirely and solely relying on fee rate estimates from a fee rate forecaster.

murchandamus commented on src/wallet/rpc/spend.cpp:
```diff
@@ -456,6 +460,53 @@ RPCHelpMan settxfee()
+        [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
+{
+
+    CAmount test = AmountFromValue(request.params[0]);
+    double conv_ammount = (static_cast<double>(test) / COIN) / 100000;  // Convert to BTC/kvB
+    CAmount nAmount = AmountFromValue(UniValue(conv_ammount));
```
Also the variable names here are a bit odd: "test" and "conv_ammount" with two "m"?

polespinasa replied:
Thanks for the reviews! As we are not going to introduce settxfeerate these corrections will not be necessary.

murchandamus commented on src/wallet/rpc/spend.cpp:
```diff
@@ -456,6 +460,53 @@ RPCHelpMan settxfee()
+    if (tx_fee_rate == CFeeRate(0)) {
+        // automatic selection
+    } else if (tx_fee_rate < pwallet->chain().relayMinFee()) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("txfee cannot be less than min relay tx fee (%s)", pwallet->chain().relayMinFee().ToString(FeeEstimateMode::SAT_VB)));
+    } else if (tx_fee_rate < pwallet->m_min_fee) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("txfee cannot be less than wallet min fee (%s)", pwallet->m_min_fee.ToString(FeeEstimateMode::SAT_VB)));
+    } else if (tx_fee_rate > max_tx_fee_rate) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("txfee cannot be more than wallet max tx fee (%s)", max_tx_fee_rate.ToString(FeeEstimateMode::SAT_VB)));
```
Let’s distinguish between fees and feerates: instead of "txfee" these errors should probably say "tx fee rate"

polespinasa replied:
Thx for the review @murchandamus. This keyword (txfee) is also used in `settxfee` function. Do you think it should be changed there also?
https://github.com/bitcoin/bitcoin/blob/0a931a9787b196d7a620863cc143d9319ffd356d/src/wallet/rpc/spend.cpp#L445-L451

murchandamus replied:
Unfortunately, there are feerates titled "…fee" all over the codebase. If you are touching the prospective lines already, you could fix the terminology, but if we are going to remove `settxfee` anyway, it’s not necessary to change it first and then remove it.

murchandamus reviewed (commented):
I concur with @ismaelsadeeq, do we think that anyone is still using `settxfee`? Should we even introduce `settxfeerate`?

polespinasa commented:
> I concur with @ismaelsadeeq, do we think that anyone is still using `settxfee`? Should we even introduce `settxfeerate`?

I agree and it makes sense not to introduce `settxfeerate` dynamic fee environment make it obsolete.

Although core estimations are not really accurate right now to rely on them. I think @ismaelsadeeq proposed an improvement on Delving Bitcoin.

fjahr commented on test/functional/wallet_create_tx.py:
```diff
@@ -23,6 +23,9 @@ def add_options(self, parser):
         self.num_nodes = 1
+        self.extra_args = [[
+            "-deprecatedrpc=settxfee"
+        ] for i in range(self.num_nodes)]
```
nit: There is just one node here, so no need for such a loop

fjahr commented on src/wallet/rpc/spend.cpp:
```diff
@@ -418,7 +418,7 @@ RPCHelpMan sendmany()
 {
     return RPCHelpMan{"settxfee",
-                "\nSet the transaction fee rate in " + CURRENCY_UNIT + "/kvB for this wallet. Overrides the global -paytxfee command line parameter.\n"
+                "\n(DEPRECATED)Set the transaction fee rate in " + CURRENCY_UNIT + "/kvB for this wallet. Overrides the global -paytxfee command line parameter.\n"
```
nit
```suggestion
                "\n(DEPRECATED) Set the transaction fee rate in " + CURRENCY_UNIT + "/kvB for this wallet. Overrides the global -paytxfee command line parameter.\n"
```

fjahr reviewed (commented):
You should squash the two commits so that `settxfeerate` is never created in the first place. That will make review much easier.

fjahr reviewed (commented):
tACK a0ad6ec5d7b3e0cd337c03a5c8a27b87b928d1c1

FYI, it's common to use prefixes for the modules touched in the commit too, so you could add `rpc:` or `wallet, rpc:` to the commit title if you retouch.

polespinasa commented on src/wallet/rpc/spend.cpp:
```diff
@@ -437,6 +437,11 @@ RPCHelpMan settxfee()
 
+    if (!pwallet->chain().rpcEnableDeprecated("settxfee")) {
+        throw JSONRPCError(RPC_METHOD_DEPRECATED, "settxfee is deprecated and will be fully removed in v30.0."
+        "\n To use settxfee in v29.0, restart bitcoind with -deprecatedrpc=settxfee.");
```
It can be deleted yes. I added it as in the example mentioned before is used: https://github.com/bitcoin/bitcoin/commit/d4cdbd6fb6ac3663d069307c4fd0078f4ecf0245#diff-ccc24453c13307f815879738d3bf00eec351417537fbf10dde1468180cacd2f1R794

fjahr commented:
reACK b9060de308fd2691cd8ac82821fe2cd5a20cd948

ismaelsadeeq reviewed (commented):
utACK b9060de308fd2691cd8ac82821fe2cd5a20cd948

ismaelsadeeq commented:
@polespinasa 
nit:
Update the PR description to reflect the recent changes

polespinasa commented:
> the reason should be clearly communicated to reviewers and documented for users.

Thanks for the comments. Updated the PR description with all the changes and idea proposed during the revisions.
Also added a release note.

murchandamus commented on doc/release-notes-31278.md:
```diff
@@ -0,0 +1,5 @@
+---
+- The `settxfee` RPC is now **deprecated** and will be removed in Bitcoin Core 30.0. This RPC allowed users to set a static fee rate for wallet transactions, but in a dynamic fee environment, this can lead to overpaying or underpaying. Users should instead rely on fee estimation or specify a fee rate per transaction. (#31278)
+
+As a reminder, manual fee rate selection should be done per transaction rather than globally. Fee estimation provides a more adaptive approach, and users can override it on a per-transaction basis when necessary.
```
The last line here mostly repeats the previous sentence. Perhaps this could be deduplicated.

murchandamus commented on test/functional/wallet_multiwallet.py:
```diff
@@ -46,7 +46,7 @@ def set_test_params(self):
         self.num_nodes = 2
         self.rpc_timeout = 120
-        self.extra_args = [["-nowallet"], []]
+        self.extra_args = [["-nowallet", "-deprecatedrpc=settxfee"], ["-deprecatedrpc=settxfee"]]
```
Why does `deprecatedrpc=settxfee` need to be specified twice here? Especially on a node without a wallet, the setting doesn’t seem to make sense.

polespinasa replied:
Actually the `-nowallet` ones needs it!! But the second one does not! Thanks for pointing this!

murchandamus reviewed (commented):
Concept ACK

polespinasa commented:
> * I expect some users not using Bitcoin Core's fee estimation, but an external one. This means that the fee estimates will need to be fed to Bitcoin Core. One way to do it is via the global RPC, another one is via the per-transaction RPC(s).

If a user relies on an external fee estimation service, setting a static wallet-wide fee rate makes little sense. Fees are dynamic, and the best practice is to set them individually per transaction rather than globally regardless of what you use to estimate them.

> * If a user wants to shoot themselves into the foot by setting the feerate to a constant, they can also do so via the per-transaction RPCs.

While users can still set a static fee manually via per-transaction RPCs, removing settxfee forces them to make that choice explicitly for each transaction, rather than applying it across all transactions by default.

This helps prevent users from accidentally sticking with an outdated fee rate, which could lead to overpayment or underpayment, increasing the risk of delayed confirmations, or the need for RBF (which harms privacy).

> So it seems like this is closing a valid use case without reducing risks or footguns.

If a user’s external software forces a static fee rate, the risk remains, but it is now a conscious decision outside of Bitcoin Core, rather than a built-in risk within the wallet itself.
By removing this option at the wallet level, we reduce the likelihood of users making costly mistakes directly in Core.

glozow commented on doc/release-notes-31278.md:
```diff
@@ -0,0 +1,3 @@
+Updated RPCs
+---
+- The `settxfee` RPC is now **deprecated** and will be removed in Bitcoin Core 30.0. This RPC allowed users to set a static fee rate for wallet transactions, but in a dynamic fee environment, this can lead to overpaying or underpaying. Users should instead rely on fee estimation or specify a fee rate per transaction. (#31278)
```
RPC examples for "specify a feerate per transaction" would be helpful

glozow reviewed (commented):
concept ACK, I wish I saw this earlier

achow101 commented:
Concept ACK

The startup option should also be deprecated.

polespinasa commented:
Thanks for reviewing. I agree, startup option should be also deprecated.
Will work on that.

maflcko reviewed (commented):
Please squash your commits according to https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits

murchandamus commented on doc/release-notes-31278.md:
```diff
@@ -0,0 +1,7 @@
+Updated RPCs
+---
+- The `settxfee` RPC is now **deprecated** and will be removed in Bitcoin Core 31.0. This RPC allowed users to set a static fee rate for wallet transactions, which is discouraged in a dynamic fee environment as it can lead to overpaying or underpaying. Users should instead rely on fee estimation or specify a fee rate per transaction using the `fee_rate` argument in RPCs like `fundrawtransaction`, ...
```
The RPC is "sendmany" without the underscore.
```suggestion
- The `settxfee` RPC is now **deprecated** and will be removed in Bitcoin Core 31.0. This RPC allowed users to set a static fee rate for wallet transactions, which is discouraged in a dynamic fee environment as it can lead to overpaying or underpaying. Users should instead rely on fee estimation or specify a fee rate per transaction using the `fee_rate` argument in RPCs like `fundrawtransaction`, `sendtoaddress`, `send`, `sendall`, and `sendmany`. (#31278)
```

polespinasa replied:
You're right, muscle memory 😅
Thanks!

murchandamus commented on doc/release-notes-31278.md:
```diff
@@ -0,0 +1,7 @@
+
+Updated settings
+------
+- The `-paytxfee` setting is now **deprecated** and will be removed in Bitcoin Core 31.0. This option set a static fee rate for all wallet transactions when starting the Bitcoin node, which is discouraged in a dynamic fee environment as it can lead to overpaying or underpaying. Users should instead rely on fee estimation or specify a fee rate per transaction using the `fee_rate` argument in RPCs  ...
```
Same:
```suggestion
- The `-paytxfee` setting is now **deprecated** and will be removed in Bitcoin Core 31.0. This option set a static fee rate for all wallet transactions when starting the Bitcoin node, which is discouraged in a dynamic fee environment as it can lead to overpaying or underpaying. Users should instead rely on fee estimation or specify a fee rate per transaction using the `fee_rate` argument in RPCs like `fundrawtransaction`, `sendtoaddress`, `send`, `sendall`, and `sendmany`. (#31278)
```

ismaelsadeeq commented on src/wallet/wallet.cpp:
```diff
@@ -3161,6 +3161,8 @@ std::shared_ptr<CWallet> CWallet::Create(WalletContext& context, const std::stri
     }
 
     if (args.IsArgSet("-paytxfee")) {
+        warnings.push_back(_("-paytxfee is deprecated and will be fully removed in v31.0."));
+
```
In "wallet,rpc: paytxfee and settxfee is deprecated and hidden from help" ee080ce689bfc8b74438a02e1c62bbe644893ffa

Shouldn't this be a failure instead of a warning since an error is being thrown in the RPC? I think we should be consistent and fail as well. 

The user should explicitly provide something like `deprecatedstartupoption=paytxfee`. Do we support that?

polespinasa replied:
I don't think we have something like `deprecatedstartupoption`.
I did it with a warn trying to follow the way it's done with Legacy wallets.
 https://github.com/bitcoin/bitcoin/blob/79bbb381a1fd13ad456fde736b3c195a791d4e58/src/wallet/wallet.cpp#L484-L487

ismaelsadeeq commented on doc/release-notes-31278.md:
```diff
@@ -0,0 +1,7 @@
+Updated RPCs
```
In "wallet,rpc: paytxfee and settxfee is deprecated and hidden from help" ee080ce689bfc8b74438a02e1c62bbe644893ffa

nit:
I think release notes is usually on it's own commit

ismaelsadeeq reviewed (commented):
Looking good @polespinasa 
I've reviewed the code and tested the RPC's and the startup option
Just a couple of comments

fjahr commented on doc/release-notes-31278.md:
```diff
@@ -1,7 +1,4 @@
-------
-- The `-paytxfee` setting is now **deprecated** and will be removed in Bitcoin Core 31.0. This option set a static fee rate for all wallet transactions when starting the Bitcoin node, which is discouraged in a dynamic fee environment as it can lead to overpaying or underpaying. Users should instead rely on fee estimation or specify a fee rate per transaction using the `fee_rate` argument in RPCs  ...
\ No newline at end of file
+RPC and Startup Option
```
You should remove the old version of the release notes from the first commit too.

polespinasa replied:
Sorry, what do you mean? There's only one commit.

fjahr replied:
Squashing is also fine! I just thought you intended to keep them in a separate commit like suggested by @ismaelsadeeq here: https://github.com/bitcoin/bitcoin/pull/31278#discussion_r1977593279

polespinasa replied:
Done :)

polespinasa commented on doc/release-notes-31278.md:
```diff
@@ -0,0 +1,7 @@
+Updated RPCs
```
added to the PR description to have that in mind

ismaelsadeeq replied:
Can also leave unresolved

polespinasa commented on src/wallet/init.cpp:
```diff
@@ -66,7 +66,7 @@ void WalletInit::AddWalletOptions(ArgsManager& argsman) const
         CURRENCY_UNIT, FormatMoney(DEFAULT_TRANSACTION_MAXFEE)), ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
```
I think this is outdated as we agreed to not hide anything?

fjahr commented on doc/release-notes-31278.md:
```diff
@@ -0,0 +1,3 @@
+RPC and Startup Option
+---
+The -paytxfee startup option and the settxfee RPC are now deprecated and will be removed in Bitcoin Core 31.0. They allowed users to set a static fee rate for wallet transactions, which is discouraged in a dynamic fee environment as it can lead to overpaying or underpaying. Users should instead rely on fee estimation or specify a fee rate per transaction using the fee_rate argument in RPCs such a ...
```
formatting nit (no need to fix unless you repush, can be done in editing of the release notes in the wiki)

``` suggestion
The `-paytxfee` startup option and the `settxfee` RPC are now deprecated and will be removed in Bitcoin Core 31.0. They allowed users to set a static fee rate for wallet transactions, which is discouraged in a dynamic fee environment as it can lead to overpaying or underpaying. Users should instead rely on fee estimation or specify a fee rate per transaction using the fee_rate argument in RPCs such as `fundrawtransaction`, `sendtoaddress`, `send`, `sendall`, and `sendmany`. (#31278)
```

jonatack replied:
also s/fee_rate/"fee_rate"/

"discouraged in a dynamic fee environment" -> this leads to the question whether the fee environment is more "dynamic" than before, e.g. when this RPC was introduced, to justify removal. Would be shorter and possibly preferable to write: "They allowed users to set a static fee rate for wallet transactions, which could potentially lead to overpaying or underpaying."

fjahr reviewed (commented):
Looks good to me aside from the missing functional deprecation test.

fjahr commented on src/wallet/rpc/spend.cpp:
```diff
@@ -437,6 +437,11 @@ RPCHelpMan settxfee()
     LOCK(pwallet->cs_wallet);
 
+    if (!pwallet->chain().rpcEnableDeprecated("settxfee")) {
+        throw JSONRPCError(RPC_METHOD_DEPRECATED, "settxfee is deprecated and will be fully removed in v31.0."
```
Note that I think it's enough to test the deprecation error there, see #31977

polespinasa committed:
Release notes

polespinasa committed:
wallet, rpc: deprecate settxfee and paytxfee

fjahr commented:
utACK f4344220d7195324f921dcf001c1a117008477fd

ismaelsadeeq commented on test/functional/rpc_deprecated.py:
```diff
@@ -4,12 +4,19 @@
 # file COPYING or http://www.opensource.org/licenses/mit-license.php.
 """Test deprecation of RPC calls."""
```
nit: Copyright update

polespinasa replied:
you mean the year?

ismaelsadeeq replied:
yes

ismaelsadeeq reviewed (commented):
Code review ACK f4344220d7195324f921dcf001c1a117008477fd


Nit:

>  No immediate breakage in 30.0, but settxfee and paytxfee will be removed in 30.0.
 
 No immediate breakage in 30.0, but settxfee and paytxfee will be removed in 31.0.

The release usually come as the last commit.

ismaelsadeeq commented:
re-ACK 8ceaecc89e20129f9c11727e4c7795633cfcdd2e [08477fd..8ceaec](https://github.com/bitcoin/bitcoin/compare/f4344220d7195324f921dcf001c1a117008477fd..8ceaecc89e20129f9c11727e4c7795633cfcdd2e)

fjahr commented:
re-ACK 8ceaecc89e20129f9c11727e4c7795633cfcdd2e

polespinasa commented:
Code rebased to avoid conflicts with https://github.com/bitcoin/bitcoin/pull/31977

fjahr commented:
re-ACK 2f2ab47bf74f4da37aad75a186cb0bb16e8af579

ismaelsadeeq commented:
re-ACK 2f2ab47bf74f4da37aad75a186cb0bb16e8af579

rkrux commented on test/functional/rpc_deprecated.py:
```diff
@@ -23,7 +31,8 @@ def run_test(self):
         # at least one other functional test that still tests the RPCs
         # functionality using the respective -deprecatedrpc flag.
 
-        self.log.info("Currently no tests for deprecated RPC methods")
```
Having reviewed this PR https://github.com/bitcoin/bitcoin/pull/31977 earlier, I feel ideally this log should not be removed but commented at this stage and the above comment should also generally mention to uncomment this log later when the RPCs tested below are fully removed from the codebase.
But we can get this done in a separate PR now after this one is merged.

polespinasa replied:
I agree. If it is not done in a separate PR we can simply add it back into the PR where we remove the RPC call. We will have to leave the test to its default form again anyway.

rkrux reviewed (commented):
Concept and utACK 2f2ab47bf74f4da37aad75a186cb0bb16e8af579

I am in favour of having both a consistent fee rate unit to specify across RPCs (sat/vB over BTC/kvB), and relying on fee estimation most of the times while creating the transaction that the user can override by setting the `fee_rate` argument in the transaction creation RPCs (including PSBT ones), which is expressed in `sat/vB`.

As a user, deprecating this RPC and startup option streamlines the overall flows for me by not having to navigate through multiple arguments & defaults while making it easier for me to gauge the fee rates used in the transaction creation.

polespinasa commented on test/functional/rpc_deprecated.py:
```diff
@@ -4,13 +4,21 @@
+from test_framework.util import assert_raises_rpc_error
 
 class DeprecatedRpcTest(BitcoinTestFramework):
+    def add_options(self, parser):
+        self.add_wallet_options(parser)
+
     def set_test_params(self):
         self.num_nodes = 1
         self.setup_clean_chain = True
         self.extra_args = [[]]
 
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+
```
noted! I will remove it anyway when removing the rpc call

polespinasa replied:
Make sense, will do. One question, as this is already merged, should this change be squashed with the other commits, should they be added on top of them or should be commited in another PR?

polespinasa replied:
noted, will do

polespinasa replied:
@maflcko please see https://github.com/bitcoin/bitcoin/pull/32139
